File: app/src/main/java/com/example/myapplication/SamplingActivity.java
package com.example.myapplication;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.snackbar.Snackbar;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.documentfile.provider.DocumentFile;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.data.AssetRepository;
import com.example.myapplication.data.CsvFolder;
import com.example.myapplication.data.CsvManager;
import com.example.myapplication.logic.ScanClassifier;
import com.example.myapplication.logic.ScannedTag;
import com.example.myapplication.model.Asset;
import com.example.myapplication.ui.AssetAdapter;

import java.util.ArrayList;
import java.util.List;

/**
 * 抽盤主畫面：與全盤一致的財產列表（可篩選）。點任一列 → 以該筆為目標進入抽盤掃描，
 * 掃描相符／確認不符後回寫並即時更新列表。
 */
public class SamplingActivity extends AppCompatActivity {

    private static final String TAG = "SamplingActivity";

    private List<Asset> assets;
    private final List<Asset> filteredAssets = new ArrayList<>();
    private AssetAdapter adapter;

    private enum Filter { ALL, UNCHECKED, MATCHED, UNMATCHED }
    private Filter currentFilter = Filter.ALL;

    private TextView tvProgress, tvEmpty;
    private TextView tabAll, tabUnchecked, tabMatched, tabUnmatched;
    private View tabIndicator;
    private ProgressBar progressLoading;
    private RecyclerView recyclerView;

    // ── 掃描啟動器 ──────────────────────────────────────
    private final ActivityResultLauncher<Intent> scanLauncher =
            registerForActivityResult(
                    new ActivityResultContracts.StartActivityForResult(),
                    result -> {
                        if (result.getResultCode() == Activity.RESULT_OK
                                && result.getData() != null) {
                            String raw = result.getData()
                                    .getStringExtra(SamplingScanActivity.RESULT_RAW);
                            if (raw != null) handleScanResult(raw);
                        }
                    }
            );

    // ── 資料夾選擇器 ────────────────────────────────────
    private final ActivityResultLauncher<Uri> folderPicker =
            registerForActivityResult(
                    new ActivityResultContracts.OpenDocumentTree(),
                    treeUri -> {
                        if (treeUri != null) onFolderPicked(treeUri);
                    }
            );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sampling);

        ViewCompat.setOnApplyWindowInsetsListener(
                findViewById(R.id.sampling_root),
                (view, insets) -> {
                    Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                    view.setPadding(0, bars.top, 0, bars.bottom);
                    return WindowInsetsCompat.CONSUMED;
                }
        );

        tvProgress   = findViewById(R.id.tv_progress);
        tvEmpty      = findViewById(R.id.tv_empty);
        progressLoading = findViewById(R.id.progress_loading);
        recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        tabAll       = findViewById(R.id.tab_all);
        tabUnchecked = findViewById(R.id.tab_unchecked);
        tabMatched   = findViewById(R.id.tab_matched);
        tabUnmatched = findViewById(R.id.tab_unmatched);
        tabIndicator = findViewById(R.id.tab_indicator);

        findViewById(R.id.btn_load).setOnClickListener(v -> folderPicker.launch(null));

        tabAll.setOnClickListener(v -> selectFilter(Filter.ALL, tabAll));
        tabUnchecked.setOnClickListener(v -> selectFilter(Filter.UNCHECKED, tabUnchecked));
        tabMatched.setOnClickListener(v -> selectFilter(Filter.MATCHED, tabMatched));
        tabUnmatched.setOnClickListener(v -> selectFilter(Filter.UNMATCHED, tabUnmatched));

        // 若記憶體已有清單（例如從其他頁返回），直接顯示
        List<Asset> existing = AssetRepository.getInstance().getAssets();
        if (existing != null && !existing.isEmpty()) {
            assets = existing;
            bindAdapter();
            selectFilter(Filter.ALL, tabAll);
            updateProgress();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (adapter != null) {
            refreshList();
            updateProgress();
        }
    }

    // 選了資料夾：取得持久權限、找出資料夾內的 CSV（多個則讓使用者選）
    private void onFolderPicked(Uri treeUri) {
        getContentResolver().takePersistableUriPermission(
                treeUri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION |
                        Intent.FLAG_GRANT_WRITE_URI_PERMISSION
        );
        AssetRepository.getInstance().setTreeUri(treeUri);

        List<DocumentFile> csvs = CsvFolder.findCsvFiles(this, treeUri);
        if (csvs.isEmpty()) {
            Toast.makeText(this, "此資料夾內找不到 CSV 檔", Toast.LENGTH_LONG).show();
            return;
        }
        if (csvs.size() == 1) {
            confirmThenUse(csvs.get(0).getUri());
            return;
        }
        String[] names = new String[csvs.size()];
        for (int i = 0; i < csvs.size(); i++) names[i] = csvs.get(i).getName();
        new AlertDialog.Builder(this)
                .setTitle("選擇 CSV 檔")
                .setItems(names, (d, which) -> confirmThenUse(csvs.get(which).getUri()))
                .show();
    }

    // 有進行中的盤點進度時，載入新清單前先警示（避免無聲覆蓋未匯出的進度）
    private boolean hasProgress() {
        if (assets == null) return false;
        for (Asset a : assets) {
            if (a.status != Asset.Status.UNCHECKED) return true;
        }
        return false;
    }

    private void confirmThenUse(Uri csvUri) {
        if (!hasProgress()) {
            useCsv(csvUri);
            return;
        }
        long total   = assets.size();
        long checked = assets.stream().filter(a -> a.status != Asset.Status.UNCHECKED).count();
        new AlertDialog.Builder(this)
                .setTitle("載入新清單")
                .setMessage("目前已有進行中的盤點（已盤 " + checked + " / " + total
                        + " 筆）。載入新清單會取代目前資料，未匯出的進度將不保留。確定載入？")
                .setPositiveButton("載入新清單", (d, w) -> useCsv(csvUri))
                .setNegativeButton("繼續原作業", null)
                .show();
    }

    private void useCsv(Uri csvUri) {
        AssetRepository.getInstance().setCsvUri(csvUri);
        loadCsv(csvUri);
    }

    // ── 載入 CSV ─────────────────────────────────────────
    private void loadCsv(Uri uri) {
        progressLoading.setVisibility(View.VISIBLE);
        new Thread(() -> {
            try {
                List<Asset> result = CsvManager.read(getContentResolver(), uri);
                runOnUiThread(() -> {
                    progressLoading.setVisibility(View.GONE);
                    if (result.isEmpty()) {
                        showError("CSV 沒有可用資料，請確認格式（需含 財產編號／名稱／部門／地點 欄位）");
                        return;
                    }
                    assets = result;
                    AssetRepository.getInstance().setAssets(assets);
                    bindAdapter();
                    selectFilter(Filter.ALL, tabAll);
                    updateProgress();
                    Snackbar.make(findViewById(R.id.sampling_root),
                            "已載入 " + assets.size() + " 筆財產",
                            Snackbar.LENGTH_LONG).show();
                });
            } catch (Exception e) {
                Log.e(TAG, "讀取失敗", e);
                runOnUiThread(() -> {
                    progressLoading.setVisibility(View.GONE);
                    showError("無法讀取 CSV，請確認檔案未損毀且為 Big5 編碼");
                });
            }
        }).start();
    }

    private void showError(String msg) {
        Snackbar.make(findViewById(R.id.sampling_root), msg, Snackbar.LENGTH_LONG).show();
    }

    private void bindAdapter() {
        adapter = new AssetAdapter(this, filteredAssets);
        adapter.setOnAssetClickListener(this::launchScan);
        recyclerView.setAdapter(adapter);
    }

    // ── 點列 → 以該筆為目標進入抽盤掃描 ──────────────────
    private void launchScan(Asset target) {
        Intent intent = new Intent(this, SamplingScanActivity.class);
        intent.putExtra(SamplingScanActivity.EXTRA_TARGET_ID, target.id);
        intent.putExtra(SamplingScanActivity.EXTRA_TARGET_NAME, target.name);
        scanLauncher.launch(intent);
    }

    // ── 處理掃描結果（回寫 + 更新列表）───────────────────
    private void handleScanResult(String raw) {
        if (assets == null) return;
        ScannedTag tag = ScannedTag.parse(raw);

        Asset target = null;
        for (Asset a : assets) {
            if (a.id.equals(tag.id)) { target = a; break; }
        }
        if (target == null) return;

        boolean isMatched = ScanClassifier.matches(target, tag);
        AssetRepository.getInstance().recordCheck(target,
                isMatched ? Asset.Status.MATCHED : Asset.Status.UNMATCHED);

        refreshList();
        updateProgress();
        Toast.makeText(this,
                isMatched ? "✅ 盤點成功（相符）" : "⚠️ 盤點完成（不相符）",
                Toast.LENGTH_SHORT).show();
    }

    // ── 篩選 / 清單 ──────────────────────────────────────
    private void selectFilter(Filter filter, TextView tab) {
        currentFilter = filter;
        moveIndicatorTo(tab);
        refreshList();
    }

    private void moveIndicatorTo(TextView tab) {
        int active   = ContextCompat.getColor(this, R.color.evergreen_primary);
        int inactive = ContextCompat.getColor(this, R.color.text_secondary);
        tabAll.setTextColor(tab == tabAll ? active : inactive);
        tabUnchecked.setTextColor(tab == tabUnchecked ? active : inactive);
        tabMatched.setTextColor(tab == tabMatched ? active : inactive);
        tabUnmatched.setTextColor(tab == tabUnmatched ? active : inactive);

        // 選中態填色（品牌綠 soft），其餘透明
        styleTabFill(tabAll, tab == tabAll);
        styleTabFill(tabUnchecked, tab == tabUnchecked);
        styleTabFill(tabMatched, tab == tabMatched);
        styleTabFill(tabUnmatched, tab == tabUnmatched);

        tab.post(() -> {
            ViewGroup.LayoutParams lp = tabIndicator.getLayoutParams();
            lp.width = tab.getWidth();
            tabIndicator.setLayoutParams(lp);
            tabIndicator.setX(tab.getX());
        });
    }

    private void styleTabFill(TextView tab, boolean active) {
        if (active) {
            tab.setBackgroundResource(R.drawable.bg_tab_active);
        } else {
            tab.setBackgroundColor(android.graphics.Color.TRANSPARENT);
        }
    }

    private void refreshList() {
        if (assets == null) return;
        filteredAssets.clear();
        for (Asset a : assets) {
            boolean include;
            switch (currentFilter) {
                case UNCHECKED: include = a.status == Asset.Status.UNCHECKED; break;
                case MATCHED:   include = a.status == Asset.Status.MATCHED;   break;
                case UNMATCHED: include = a.status == Asset.Status.UNMATCHED; break;
                case ALL:
                default:        include = true;
            }
            if (include) filteredAssets.add(a);
        }
        if (adapter != null) adapter.notifyDataSetChanged();

        if (filteredAssets.isEmpty()) {
            tvEmpty.setVisibility(View.VISIBLE);
            tvEmpty.setText(getEmptyText());
        } else {
            tvEmpty.setVisibility(View.GONE);
        }
    }

    private String getEmptyText() {
        switch (currentFilter) {
            case UNCHECKED: return "沒有未盤點的資產";
            case MATCHED:   return "沒有已盤點的資產";
            case UNMATCHED: return "沒有不相符的資產";
            default:        return "尚未載入資料";
        }
    }

    private void updateProgress() {
        if (assets == null) return;
        int total = assets.size();
        long unchecked = assets.stream().filter(a -> a.status == Asset.Status.UNCHECKED).count();
        long matched   = assets.stream().filter(a -> a.status == Asset.Status.MATCHED).count();
        long unmatched = assets.stream().filter(a -> a.status == Asset.Status.UNMATCHED).count();
        long checked   = matched + unmatched;

        tvProgress.setText("進度：" + checked + " / " + total);
        tabAll.setText("全部 " + total);
        tabUnchecked.setText("未盤點 " + unchecked);
        tabMatched.setText("已盤點 " + matched);
        tabUnmatched.setText("不相符 " + unmatched);
    }

    // ── 定點寫檔 ─────────────────────────────────────────
    private void flushCsvAsync() {
        new Thread(() -> {
            try {
                AssetRepository.getInstance().flush(getContentResolver());
            } catch (Exception e) {
                runOnUiThread(() ->
                        Toast.makeText(this, "CSV 寫入失敗：" + e.getMessage(),
                                Toast.LENGTH_LONG).show());
                Log.e(TAG, "CSV 寫入失敗", e);
            }
        }).start();
    }

    @Override
    protected void onPause() {
        super.onPause();
        flushCsvAsync();
    }

    @Override
    protected void onStop() {
        super.onStop();
        flushCsvAsync();
    }
}

File: app/src/main/java/com/example/myapplication/SamplingActivity.kt
package com.example.myapplication

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.example.myapplication.data.AssetRepository
import com.example.myapplication.data.CsvManager
import com.example.myapplication.model.Asset
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * 抽盤頁：逐筆瀏覽財產，對單一目標啟動掃描比對。
 *
 * Kotlin 筆記：
 * - 只讀取清單，故 `assets` 用唯讀的 `List<Asset>?`（與 ScanActivity 的 MutableList 對照）。
 * - `return@setOnClickListener` / `return@runOnUiThread`：從 lambda 中提前返回要標記回哪個 lambda。
 * - `assets.any { }` / `count { }` 取代 Java Stream。
 */
class SamplingActivity : AppCompatActivity() {

    private var assets: List<Asset>? = null
    private var currentIndex = 0 // 目前顯示第幾筆

    // ── DOM ──────────────────────────────────────────────
    private lateinit var tvProgress: TextView
    private lateinit var tvPosition: TextView
    private lateinit var tvResult: TextView
    private lateinit var tvId: TextView
    private lateinit var tvName: TextView
    private lateinit var tvDepartment: TextView
    private lateinit var tvLocation: TextView
    private lateinit var btnLoad: Button
    private lateinit var btnPrev: Button
    private lateinit var btnNext: Button
    private lateinit var btnScan: Button
    private lateinit var btnDone: Button
    private lateinit var btnJumpPrev: Button
    private lateinit var btnJumpNext: Button

    // 掃描啟動器
    private val scanLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                val raw = result.data?.getStringExtra(SamplingScanActivity.RESULT_RAW)
                if (raw != null) {
                    handleScanResult(raw)
                }
            }
        }

    // ── 檔案選擇器 ──────────────────────────────────────
    private val filePicker =
        registerForActivityResult(ActivityResultContracts.OpenDocument()) { uri: Uri? ->
            if (uri != null) {
                contentResolver.takePersistableUriPermission(
                    uri,
                    Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION,
                )
                AssetRepository.csvUri = uri
                loadCsv(uri)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sampling)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.sampling_root)) { view, insets ->
            val bars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            view.setPadding(0, bars.top, 0, bars.bottom)
            WindowInsetsCompat.CONSUMED
        }

        bindViews()
        setupListeners()
    }

    private fun bindViews() {
        tvProgress = findViewById(R.id.tv_progress)
        tvPosition = findViewById(R.id.tv_position)
        tvResult = findViewById(R.id.tv_result)
        tvId = findViewById(R.id.tv_id)
        tvName = findViewById(R.id.tv_name)
        tvDepartment = findViewById(R.id.tv_department)
        tvLocation = findViewById(R.id.tv_location)
        btnLoad = findViewById(R.id.btn_load)
        btnPrev = findViewById(R.id.btn_prev)
        btnNext = findViewById(R.id.btn_next)
        btnScan = findViewById(R.id.btn_scan)
        btnDone = findViewById(R.id.btn_done)
        btnJumpPrev = findViewById(R.id.btn_jump_prev)
        btnJumpNext = findViewById(R.id.btn_jump_next)
    }

    private fun setupListeners() {
        btnLoad.setOnClickListener { filePicker.launch(arrayOf("*/*")) }

        btnPrev.setOnClickListener { showAssetAt(currentIndex - 1) }
        btnNext.setOnClickListener { showAssetAt(currentIndex + 1) }

        btnJumpPrev.setOnClickListener { jumpToUnchecked(false) }
        btnJumpNext.setOnClickListener { jumpToUnchecked(true) }

        btnScan.setOnClickListener {
            val assets = assets ?: return@setOnClickListener
            if (assets.isEmpty()) return@setOnClickListener
            val target = assets[currentIndex]

            val intent = Intent(this, SamplingScanActivity::class.java)
            intent.putExtra(SamplingScanActivity.EXTRA_TARGET_ID, target.id)
            intent.putExtra(SamplingScanActivity.EXTRA_TARGET_NAME, target.name)
            scanLauncher.launch(intent)
        }

        btnDone.setOnClickListener { finish() }
    }

    // ── 載入 CSV ─────────────────────────────────────────
    private fun loadCsv(uri: Uri) {
        Thread {
            try {
                val result = CsvManager.read(contentResolver, uri)
                runOnUiThread {
                    assets = result
                    AssetRepository.assets = result

                    if (result.isEmpty()) {
                        Toast.makeText(this, "CSV 是空的", Toast.LENGTH_SHORT).show()
                        return@runOnUiThread
                    }

                    currentIndex = 0
                    showAssetAt(0)
                    enableActionButtons(true)
                    Toast.makeText(this, "載入成功：${result.size} 筆", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                runOnUiThread {
                    Toast.makeText(this, "讀取失敗：${e.message}", Toast.LENGTH_LONG).show()
                }
                Log.e(TAG, "讀取失敗", e)
            }
        }.start()
    }

    // ── 顯示指定位置的資產 ───────────────────────────────
    private fun showAssetAt(index: Int) {
        val assets = assets ?: return
        if (assets.isEmpty()) return
        if (index < 0 || index >= assets.size) return

        currentIndex = index
        val a = assets[index]

        tvId.text = a.id
        tvName.text = a.name
        tvDepartment.text = a.department.ifEmpty { "－" }
        tvLocation.text = a.location.ifEmpty { "－" }

        tvPosition.text = "第 ${index + 1} 筆 / 共 ${assets.size} 筆"

        // 比對結果顯示
        when (a.status) {
            Asset.Status.MATCHED -> {
                tvResult.text = "✅ 相符"
                tvResult.setTextColor(Color.parseColor("#16a34a"))
                tvResult.setBackgroundColor(Color.parseColor("#f0fdf4"))
            }
            Asset.Status.UNMATCHED -> {
                tvResult.text = "⚠️ 不相符"
                tvResult.setTextColor(Color.parseColor("#c2410c"))
                tvResult.setBackgroundColor(Color.parseColor("#fff7ed"))
            }
            Asset.Status.UNCHECKED -> {
                tvResult.text = "待掃描"
                tvResult.setTextColor(Color.parseColor("#8e8e93"))
                tvResult.setBackgroundColor(Color.parseColor("#f2f2f7"))
            }
        }

        updateNavButtons()
        updateProgress()
    }

    // ── 跳到上/下一筆未盤點 ──────────────────────────────
    private fun jumpToUnchecked(forward: Boolean) {
        val assets = assets ?: return
        val total = assets.size
        val step = if (forward) 1 else -1

        for (i in 1..total) {
            val idx = currentIndex + step * i
            if (idx < 0 || idx >= total) break // 不繞圈，到頂/到底就停

            if (assets[idx].status == Asset.Status.UNCHECKED) {
                showAssetAt(idx)
                return
            }
        }

        Toast.makeText(
            this,
            if (forward) "無下筆未盤點的資產" else "無上筆未盤點的資產",
            Toast.LENGTH_SHORT,
        ).show()
    }

    // ── 更新導覽按鈕狀態 ─────────────────────────────────
    private fun updateNavButtons() {
        val assets = assets ?: return
        btnPrev.isEnabled = currentIndex > 0
        btnNext.isEnabled = currentIndex < assets.size - 1

        // 跳轉按鈕：只要還有未盤點的就可用
        val hasUnchecked = assets.any { it.status == Asset.Status.UNCHECKED }
        btnJumpPrev.isEnabled = hasUnchecked
        btnJumpNext.isEnabled = hasUnchecked
    }

    // ── 更新進度顯示 ─────────────────────────────────────
    private fun updateProgress() {
        val assets = assets ?: return
        val checked = assets.count { it.status != Asset.Status.UNCHECKED }
        tvProgress.text = "進度：$checked / ${assets.size} 已盤點"
    }

    // ── 啟用底部操作按鈕 ─────────────────────────────────
    private fun enableActionButtons(enabled: Boolean) {
        btnScan.isEnabled = enabled
    }

    // ── 處理掃描結果 ─────────────────────────────────────
    private fun handleScanResult(raw: String) {
        val assets = assets ?: return
        if (assets.isEmpty()) return

        val target = assets[currentIndex]

        // 拆解 QR Code
        val parts = raw.split(";")
        val scannedId = parts.getOrNull(0)?.trim().orEmpty()
        val scannedDepartment = parts.getOrNull(2)?.trim().orEmpty()
        val scannedLocation = parts.getOrNull(3)?.trim().orEmpty()

        // 雙重保險：理論上 SamplingScanActivity 已經過濾掉非目標的資產
        if (scannedId != target.id) {
            Toast.makeText(this, "掃描結果與目標不符", Toast.LENGTH_SHORT).show()
            return
        }

        // 比對部門和地點
        val isMatched = target.department == scannedDepartment && target.location == scannedLocation

        target.status = if (isMatched) Asset.Status.MATCHED else Asset.Status.UNMATCHED
        target.checkedAt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(Date())

        AssetRepository.markDirty() // 只改記憶體，落檔延到 onPause/onStop

        // 重新顯示這一筆，更新 UI
        showAssetAt(currentIndex)

        Toast.makeText(
            this,
            if (isMatched) "✅ 盤點成功（相符）" else "⚠️ 盤點完成（不相符）",
            Toast.LENGTH_SHORT,
        ).show()
    }

    // ── 定點寫檔 ─────────────────────────────────────────
    // 只有記憶體有未落檔變更時才真正寫一次（背景執行緒）。
    private fun flushCsvAsync() {
        Thread {
            try {
                AssetRepository.flush(contentResolver)
            } catch (e: Exception) {
                runOnUiThread {
                    Toast.makeText(this, "CSV 寫入失敗：${e.message}", Toast.LENGTH_LONG).show()
                }
                Log.e(TAG, "CSV 寫入失敗", e)
            }
        }.start()
    }

    override fun onPause() {
        super.onPause()
        flushCsvAsync()
    }

    override fun onStop() {
        super.onStop()
        flushCsvAsync()
    }

    companion object {
        private const val TAG = "SamplingActivity"
    }
}
